### PR TITLE
Update SetupTypo3.rst

### PR DESCRIPTION
**Pull request by AKaravas**
Added needed command in order to successfully run composer install

### CODE REVIEW

**Comment by liayn** [comment]
For current master you don't need this. For other branches you need it though. I would refine the text slightly and mention this fact. What do you think?

**Comment by AKaravas** [comment]
I downloaded the master via git and I had to include it cause I got an error that it could not find which version I had

**Comment by mbrodala** [comment]
@liayn The current master doesn't have a `branch-alias` either, thus `COMPOSER_ROOT_VERSION` is required. Do you see a different behavior?

**Comment by sypets** [comment]
I never used `COMPOSER_ROOT_VERSION` for setting up core development. I am a bit surprised that this is now required.

@AKaravas Can you post your error message? Just so we have all the information.

**Comment by liayn** [comment]
@mbrodala For master I do not need this env variable. I need it for all other branches.
composer takes the master branch, if no version can be identified, afaik.

When checking out 10.4 or other versions you need to set the `COMPOSER_ROOT_VERSION` accordingly

**Comment by AKaravas** [comment]
Error

**Comment by liayn** [comment]
@AKaravas This is strange. Are you really on the master branch? I just double-checked and pulled the latest changes and ran a `composer install` with no issues.

**Comment by AKaravas** [comment]
@liayn Yes i am pretty sure. This is my git config

Git config

**Comment by mbrodala** [comment]
Hm, indeed it works just fine for me:

```
$ git clone git://git.typo3.org/Packages/TYPO3.CMS.git --depth 1 .
Cloning into '.'...
remote: Counting objects: 18119, done.
remote: Compressing objects: 100% (13283/13283), done.
remote: Total 18119 (delta 5415), reused 12500 (delta 3772)
Receiving objects: 100% (18119/18119), 27.17 MiB | 5.90 MiB/s, done.
Resolving deltas: 100% (5415/5415), done.
$ composer install
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
...
Generating autoload files
Generating empty class alias map file
composer/package-versions-deprecated: Generating version class...
composer/package-versions-deprecated: ...done generating version class
41 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
```

**Comment by sypets** [comment]
## First, to clarify why we are seeing these different results for `composer install`

@AKaravas Did you do a fresh "git clone" or did you clean up the repo and get the latest changes? And, as already pointed out made sure you are on master?

Either:
```
git checkout master
git reset --hard origin/master
git pull origin master
```
or 

`git clone` as seen in example above.

You could also show the output of the following:

```
composer --version
git br
git log -1
```

Please post results formatted as code (use the `<>` icon above) and not as screenshot. 

You error message implies, it does not know it is on 11.0.*@dev:

```
Problem 1
- Installation request for friendsoftypo3/phpstan-typo3 0.5.0 -> satisfiable by friendsoftypo3/phpstan-typo3[0.5.0].
- friendsoftypo3/phpstan-typo3 0.5.0 requires typo3/cms-core ^10.4 || 11.0.*@dev -> no matching package found
```

In any case, it would be good to find out why exactly we see this different behaviour (we are using the same repository, same version, composer.lock where the versions are already explicitly specified, so we should not have this different behaviour).

## How does Composer know typo3/cms-core is installed and which version is installed?

This might be beyond the scope of this PR but I would like to understand better how it works. 

Because of the core development setup (because we want to get and change the system extensions and not install them from packagist) the system extensions are not installed via Composer, they already exist after `git clone` (in typo3/sysext). So how does Composer know which TYPO3 (or rather typo3/cms-core) version it is dealing with when it checks the version constraints? How does it even know it is installed (esp. since it is installed in `typo3/sysext`, a directory Composer doesn't really know anything about..

(If this is too much here, I can ask the question somewhere else, or try to dig in and find out myself.)

**Comment by liayn** [comment]
> How does it even know it is installed (esp. since it is installed in typo3/sysext, a directory Composer doesn't really know anything about..

It does not need to. The root composer.json file does not require the individual sysexts and even defines them as `replace` section. See https://git.typo3.org/Packages/TYPO3.CMS.git/blob_plain/HEAD:/composer.json 
So whatever sysext is required by a package, is "covered" with the `replace` section.
Furthermore the master branch defines a branch alias as `"11.0.x-dev"`. That's where the version info is fetched from.

**Comment by mbrodala** [comment]
Indeed, no idea why I missed this:

https://github.com/TYPO3/TYPO3.CMS/blob/7a0b3c2/composer.json#L119-L121

**Comment by sypets** [comment]
> It does not need to. The root composer.json file does not require the individual sysexts and even defines them as replace section. 

Yes, I saw that. But the reason we have this PR is that version conflicts were reported in combination with phpstan-typo3. This requires a specific cms-core version and that is where the conflict was reported. So apparently, Composer does try to check if it can resolve the version constraints and for that reason it needs to know that cms-core is installed and which version. 

----

But I think you answered my question with the branch-alias.

```
"branch-alias": {
			"dev-master": "11.0.x-dev"
		}
```

in the `composer.json`

---

Thanks for clearing that up! 

---

Does that explain why the COMPOSER_ROOT_VERSION is required in the other branches? Because they do not have that (I checked for 9.5).

**Comment by liayn** [comment]
> Does that explain why the COMPOSER_ROOT_VERSION is required in the other branches? Because they do not have that (I checked for 9.5).

Yes, I noticed that too, I still need to read up the exact process composer is resolving version numbers

**Comment by linawolf** [comment]
What is the status of this PR? @liayn @mbrodala @sypets ? Should it be merged or closed?

**Comment by liayn** [comment]
Lets keep the information, but adjust it slightly. I prefer to have 
`COMPOSER_ROOT_VERSION=10.4 composer install` instead of exporting `COMPOSER_ROOT_VERSION` globally.

**Comment by mbrodala** [comment]
I'd agree on the latter. Personally I have wrapped the TYPO3 source with a custom Docker Compose setup where I define this as `environment`. Maybe we could mention something like this.

**Comment by liayn** [comment]
> 
> 
> I'd agree on the latter. Personally I have wrapped the TYPO3 source with a custom Docker Compose setup where I define this as `environment`. Maybe we could mention something like this.

Me too! So full ack on adding this as well, looks like this:

``` 
version: '3.7'

services:
  db:
    image: mysql:5.7
    command: mysqld --character-set-server=utf8mb4 --collation-server=utf8mb4_general_ci
    ports:
      - "3306:3306"
    environment:
      - MYSQL_ROOT_PASSWORD=root
      - MYSQL_DATABASE=typo3
      - MYSQL_USER=typo3
      - MYSQL_PASSWORD=passme
    volumes:
      - type: bind
        source: ./db-data
        target: /var/lib/mysql
  web:
    image: <some PHP image>
    depends_on:
      - db
    ports:
      - "80:80"
    volumes:
      - type: bind
        source: ./html
        target: /var/www/html
      - type: bind
        source: <path to git repo>
        target: /var/www/typo3_src
    environment:
      - COMPOSER_ROOT_VERSION=10.4
      - TYPO3_CONTEXT=Development
```

**Comment by linawolf** [comment]
@liayn @mbrodala can this be merged or should it be closed? If it is still work in progress please put it in draft status

File: Documentation/Setup/SetupTypo3.rst [hunk 1]
```diff
@@ -60,6 +60,12 @@ see :ref:`composer-install`)
 
 Information about :ref:`setting up Composer <prerequisites-composer>` is found in previous chapter.
 
+Composer cannot detect the TYPO3 version of your cloned project because there is none. Before you run 
+`composer install` you need to run the following::
```
Comment (mbrodala) [inline Documentation/Setup/SetupTypo3.rst:64]:
```suggestion
`composer install` you need to export the `COMPOSER_ROOT_VERSION environment variable <https://getcomposer.org/doc/03-cli.md#composer-root-version>`__.
Here you need to set a full version string matching the TYPO3 version of your clone.

Example::
```

**Review by mbrodala — APPROVED** [review]

**Pull request merged**